File: LCS.py
def LCS(X, Y):
  """
  This function calculates the Longest Common Subsequence (LCS) of two strings.

  Args:
    X: The first string.
    Y: The second string.

  Returns:
    A tuple containing the length of the LCS and the LCS itself.
  """
  n, m = len(X), len(Y)
  dp = [[0 for _ in range(m + 1)] for _ in range(n + 1)]

  for i in range(1, n + 1):
    for j in range(1, m + 1):
      if X[i - 1] == Y[j - 1]:
        dp[i][j] = dp[i - 1][j - 1] + 1
      else:
        dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

  lcs = ""
  i, j = n, m
  while i > 0 and j > 0:
    if X[i - 1] == Y[j - 1]:
      lcs += X[i - 1]
      i -= 1
      j -= 1
    else:
      if dp[i - 1][j] >= dp[i][j - 1]:
        i -= 1
      else:
        j -= 1

  return dp[n][m], lcs[::-1]

# Example usage
X = "wildfire"
Y = "fireworks"

lcs_length, lcs = LCS(X, Y)

# Calculate the string similarity score
similarity_score = (2 * lcs_length) / (len(X) + len(Y))

print(f"Length of LCS: {lcs_length}")
print(f"LCS: {lcs}")
print(f"String similarity score: {similarity_score}")

File: LCS.cpp
#include <iostream>
#include <string>
using namespace std;
int LCS(const string &X, const string &Y, int i, int j)
{
    if (i == X.length() || j == Y.length())
    {
        return 0;
    }
    if (X[i] == Y[j])
    {
        return 1 + LCS(X, Y, i + 1, j + 1);
    }
    else
    {
        return max(LCS(X, Y, i + 1, j), LCS(X, Y, i, j + 1));
    }
}
int main()
{
    string X = "wildfire";
    string Y = "fireworks";
    int lcsLength = LCS(X, Y, 0, 0);
    cout << "Length of LCS: " << lcsLength << endl;
    return 0;
}
